In [1]:
import pandas as pd
import seaborn as sns
import os
import numpy as np
import matplotlib.pyplot as plt

In [2]:
os.chdir('C:\Application of Software')

In [3]:
os.getcwd()

'C:\\Application of Software'

In [7]:
RM = pd.ExcelFile('Exhibit+2.xlsx')
RM

In [9]:
Rmodel = pd.read_excel(RM, 'Rocinante Models')
Mmodel= pd.read_excel(RM, 'Marengo Models')

In [10]:
Rmodel.head()

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0  Rocinante 1                   171.877                      6.1   
1  Rocinante 2                   139.796                      6.1   
2  Rocinante 3                   178.947                      9.9   
3  Rocinante 4                   140.022                      5.8   
4  Rocinante 5                   186.476                     10.0   

   Mileage\n(Km/ltr)  Top speed (Km/hr)  
0               15.8              168.2  
1               12.1              149.6  
2               17.0              173.4  
3               11.6              170.6  
4               17.2              175.0

In [15]:
y = Rmodel['Sales \n(in 1,000 units)']
x = Rmodel[['Price\n(in lakh rupees)','Mileage\n(Km/ltr)','Top speed (Km/hr)']]

In [16]:
from sklearn.model_selection import train_test_split

## Splitting the data into train and test set(70 :30)

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [18]:
 x.shape, X_train.shape, X_test.shape

((35, 3), (24, 3), (11, 3))

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

## Fitting it into Leniar Regression model

In [41]:
model = LinearRegression()
model.fit(X_train, y_train)


LinearRegression()

In [42]:
y_pred = model.predict(X_test)

In [24]:
from sklearn.metrics import mean_squared_error, r2_score 

In [25]:
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [26]:
print("Mean Squared Error:", mse)
print("R^2 Score:", r2)
print("Predictions:", y_pred)

Mean Squared Error: 5.216960179953158
R^2 Score: 0.9933849988504676
Predictions: [121.30136604 136.07163245 182.19996027 147.38174604 200.57199037
 169.00185098 191.89836208 138.00151921 203.2265794  207.47371754
 151.22829698]


## A lower MSE indicates a better fit of the model to the data. An MSE of approximately 5.22 suggests that, on average, the squared difference between the predicted and actual values is around 5.22 this could be considered a reasonable error.

## An R² score of 0.9934 means that about 99.34% of the variance in the target variable can be explained by the model. This suggest\ that the model fits the data very well.

In [27]:
rmse = np.sqrt(mse)

In [28]:
print("Root Mean Squared Error:", rmse)
print("Predictions:", y_pred)

Root Mean Squared Error: 2.2840665883360667
Predictions: [121.30136604 136.07163245 182.19996027 147.38174604 200.57199037
 169.00185098 191.89836208 138.00151921 203.2265794  207.47371754
 151.22829698]


## An RMSE of 2.284 indicates that, on average, the model's predictions deviate from the actual values by about 2.284 units.As the target variable ranges from 0 to 100, an RMSE of 2.284 is relatively small, suggesting good predictive accuracy.

In [29]:
y_train_pred = model.predict(X_train)  # Predictions for training set
y_test_pred = model.predict(X_test)     # Predictions for testing set

# Step 6: Evaluate the model performance
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_train = np.sqrt(mse_train)  # Calculate RMSE for training set
rmse_test = np.sqrt(mse_test)    # Calculate RMSE for testing set

# Display the results
print("Training Set Predictions:", y_train_pred)
print("Testing Set Predictions:", y_test_pred)
print("Training Set RMSE:", rmse_train)
print("Testing Set RMSE:", rmse_test)

Training Set Predictions: [204.82388355 174.36956355 182.85114018 203.97960124 175.17634813
 189.18389588 199.70284905 143.06426691 181.23497382 120.80465422
 139.4929367  198.62299338 103.65855769 204.96114924 171.14586446
 150.69539865 103.84794727 168.36881053 176.84093835 106.18135033
 120.72039279 149.06541703 107.50877295 117.54629411]
Testing Set Predictions: [121.30136604 136.07163245 182.19996027 147.38174604 200.57199037
 169.00185098 191.89836208 138.00151921 203.2265794  207.47371754
 151.22829698]
Training Set RMSE: 2.2339347338071893
Testing Set RMSE: 2.2840665883360667


## The lower value of the rmse for both test and train data indicates that the model predicts value will less deviation or difference.

In [30]:
if rmse_train < rmse_test:
    print("The model may be overfitting.")
elif rmse_train > rmse_test:
    print("The model may be underfitting.")
else:
    print("The model is performing well on both training and testing sets.")

The model may be overfitting.


## The model is slightly overfitting but there is not much of diffference in the rmse values of test and train data set. Hence it is reliable

## Splitting the data into Train and Test (80:20)

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [32]:
 x.shape, X_train.shape, X_test.shape

((35, 3), (28, 3), (7, 3))

In [33]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [34]:
y_pred = model.predict(X_test)

In [35]:
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [36]:
print("Mean Squared Error:", mse)
print("R^2 Score:", r2)
print("Predictions:", y_pred)

Mean Squared Error: 1.3078242422470585
R^2 Score: 0.9982303453375267
Predictions: [121.75169134 136.61925871 181.97910424 148.13961039 200.76798106
 169.3210536  191.13850189]


## An MSE of 1.308 suggests that, on average, the squared differences between the predicted values and the actual values are about 1.308.
## This is a relatively low error, indicating that the predictions are close to the actual values.

## An R² score of 0.9982 means that approximately 99.82% of the variance in the target variable can be explained by the model.
## This score indicating that the model fits the data very well and is likely to make accurate predictions.

In [37]:
rmse = np.sqrt(mse)

In [38]:
print("Root Mean Squared Error:", rmse)
print("Predictions:", y_pred)

Root Mean Squared Error: 1.143601435049405
Predictions: [121.75169134 136.61925871 181.97910424 148.13961039 200.76798106
 169.3210536  191.13850189]


## An RMSE of 1.144 indicates that, on average, the model's predictions deviate from the actual values by about 1.144 units.
## The target variable ranges from 0 to 100, an RMSE of 1.144 is relatively small, suggesting good predictive accuracy.

In [39]:
y_train_pred = model.predict(X_train)  # Predictions for training set
y_test_pred = model.predict(X_test)     # Predictions for testing set

# Step 6: Evaluate the model performance
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_train = np.sqrt(mse_train)  # Calculate RMSE for training set
rmse_test = np.sqrt(mse_test)    # Calculate RMSE for testing set

# Display the results
print("Training Set Predictions:", y_train_pred)
print("Testing Set Predictions:", y_test_pred)
print("Training Set RMSE:", rmse_train)
print("Testing Set RMSE:", rmse_test)

Training Set Predictions: [138.62336097 202.71811463 206.50249904 151.04175674 203.97167952
 174.00610718 182.51896311 203.50553436 175.29124796 188.58662052
 200.21229309 143.50855927 180.95596561 120.65746953 139.31072957
 198.99165469 104.12813358 204.00986728 171.592459   150.30175035
 104.23479659 168.72718348 175.91805584 106.612865   120.86683126
 148.8559326  108.14821709 118.30835214]
Testing Set Predictions: [121.75169134 136.61925871 181.97910424 148.13961039 200.76798106
 169.3210536  191.13850189]
Training Set RMSE: 2.361605466796605
Testing Set RMSE: 1.143601435049405


## The test data set has lower RMSE score than the train data set hence it shows that model has good predictive power on the new data 


In [40]:
if rmse_train < rmse_test:
    print("The model may be overfitting.")
elif rmse_train > rmse_test:
    print("The model may be underfitting.")
else:
    print("The model is performing well on both training and testing sets.")

The model may be underfitting.


## The high training RMSE suggests that the model is likely underfitting the training data